LibGit2Sharp/Filter.cs

```
@@ -0,0 +1,350 @@
+                thisStream.write = StreamWriteCallback;
+                thisStream.free = StreamFreeCallback;
+                thisPtr = Marshal.AllocHGlobal(Marshal.SizeOf(thisStream));
+                Marshal.StructureToPtr(thisStream, thisPtr, false);
+                nextPtr = git_writestream_next;
+                nextStream = new GitWriteStream();
+                Marshal.PtrToStructure(nextPtr, nextStream);
+                filterSource = FilterSource.FromNativePtr(filterSourcePtr);
+            }
+            catch (Exception exception)
+            {
+                // unexpected failures means memory clean up required
+                if (thisPtr != IntPtr.Zero)
+                {
+                    Marshal.FreeHGlobal(thisPtr);
+                    thisPtr = IntPtr.Zero;
+                }
+
+                Proxy.giterr_set_str(GitErrorCategory.Filter, exception.Message);
+                result = (int)GitErrorCode.Error;
+            }
+
+            git_writestream_out = thisPtr;
+
+            return result;
+        }
+
+        int StreamCloseCallback(IntPtr stream)
+        {
+            int result = 0;
+
+            try
+            {
+                Ensure.ArgumentNotZeroIntPtr(stream, "stream");
+                Ensure.ArgumentIsExpectedIntPtr(stream, thisPtr, "stream");
+
+                result = nextStream.close(nextPtr);
+            }
+            catch (Exception exception)
+            {
+                Proxy.giterr_set_str(GitErrorCategory.Filter, exception.Message);
+                result = (int)GitErrorCode.Error;
+            }
+
+            return result;
+        }
+
+        void StreamFreeCallback(IntPtr stream)
+        {
+            try
+            {
+                Ensure.ArgumentNotZeroIntPtr(stream, "stream");
+                Ensure.ArgumentIsExpectedIntPtr(stream, thisPtr, "stream");
+
+                Marshal.FreeHGlobal(thisPtr);
+            }
+            catch { }
```

This looks unexpected - do we really want to catch and swallow all exceptions? I would like to at least trace this, but we should decide the correct thing to do here for error handling (if it is just tracing and catching, then maybe document this reason in code).